wordpress-mobile/WordPress-iOS
# Deprecate `Deliverfile` in favor of `Fastfile`

_I decided to go ahead and implement all the code for this proposal in order to discuss it with a clear example available._

_Also notice this PR is against the `move-fastlane-secrets-outside-repo`, which itself is against `release/17.7`. That's because this PR use the App Store Connect API key path and it made sense to me to use it in the "new and proper" mode introduced in `move-fastlane-secrets-outside-repo`. As for the release branch, this automation is for the release process, so merging it here will allow me to test it IRL sooner rather than later._

I'd like to suggest we remove `Deliverfile` as the way to configure the parameters for `deliver` in favor of calling it with a dedicated lane holding all the settings.

Our current approach to updated the metadata on ASC is to call `deliver`, i.e.: `bundle exec fastlane deliver --skip_screenshots`. I don't know whether that was an intentional choice or it's a byproduct of following the Fastlane documentation, but I think that we would be better served by dropping `Deliverfile` and using Fastlane as the single source of truth.

Alongside having a single source of truth, which is always useful in terms of clarity, running the metadata step through a lane makes it consistent with the rest of the steps in the release management process. It also allows us to add additional checks as part of the process in the future, which we cannot do right now by calling `deliver` directly.

The downsides of this approach might be that using `Deliverfile` is, as far as I understand, the standard approach (if only because setting up `deliver` generates a `Deliverfile` for you). Also worth pointing out that, while it's handy to have a single source of truth, our `Fastfile` is pretty long already, so we're liable of information overload and difficulties in navigating the file.

Keen to hear what you think of this approach.

## To test

Unfortunately, the new lane that uploads metadata to ASC for both WordPress and Jetpack can't be tested end-to-end unless we actually upload the data. If we cancel the `deliver` submission for WordPress, the lane interrupts without giving us a chance to test the Jetpack call. As such, Jetpack needs to be tested in isolation.

1. Checkout this branch
2. Run `bundle exec fastlane update_metadata_on_app_store_connect`
3. Verify the `Preview.html` contains valid data. In particular, that it's not attempting to upload screenshots
4. **Cancel the upload** 🙏 `Does the Preview on path './fastlane/Preview.html' look okay for you? (y/n) n`
5. Repeat for Jetpack `bundle exec fastlane update_jetpack_metadata_on_app_store_connect`

_Notice how the description says Jetpack, not WordPress 😉_

6. **Cancel the upload** 🙏 `Does the Preview on path './fastlane/Preview.html' look okay for you? (y/n) n`

Optionally, call one or both lanes with `with_screenshots:true` to see that it will attempt to load the screenshots. Notice that doing so works only if you have the screenshots locally.

## Regression Notes
1. Potential unintended areas of impact
The metadata upload process, once the user approves and they reach App Store Connect.


2. What I did to test those areas of impact (or what existing automated tests I relied on)
See the steps above for checking as much as possible locally. If anything else is broken I, as the release train conductor, should be able to address it as part of the release finalization.

3. What automated tests I added (or what prevented me from doing so)
N.A.

---
